### ISSUE TITLE
AutoGraph and tf.function are not working for TPU

### ISSUE THREAD

**Issue by saahiluppal** [tensorflow/tensorflow#38828]
It seems that TPU only supports keras fit() function, but unable to use functions from tf.function and autographs.

Tensorflow version 2.1
python version 3.6

Issue can be reproduced in colab.

**Comment by amahendrakar** [comment]
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here.

You can also share the link of the Colab gist you are using. Thanks!

**Comment by saahiluppal** [comment]
Here's the link to colab file
https://colab.research.google.com/drive/1hJUOZ5VfqqxIp36jpfVX1tS-r5AUexj-

Error suggests that gradients and optimizer have different scope which is quite reasonable as GradientTape introduces it's own scope.

**Comment by amahendrakar** [comment]
> https://colab.research.google.com/drive/1hJUOZ5VfqqxIp36jpfVX1tS-r5AUexj-

@saahiluppal,
I am unable access this colab link. Please select 'File' -> 'Save a copy as Github Gist', and share the link of the new window. Thanks!

**Comment by saahiluppal** [comment]
Here's the link to gist.
https://gist.github.com/saahiluppal/0bf79c27a15eaf6a72b25322eae6b6aa

Thanks.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.1](https://colab.research.google.com/gist/amahendrakar/baa48b9fc7e2662d776cdb9dc65e9943/38828-2-1.ipynb) and [TF v2.2.0-rc3](https://colab.research.google.com/gist/amahendrakar/8ebb5491f3cc5a0bf0f6cd4ffe6e88b5/38828-2-2.ipynb). Please find the attached gist. Thanks!

**Comment by saeta** [comment]
Sorry @jvishnuvardhan , I am no longer working on TPUs. I'm going to re-assign it back to you to re-triage. Thanks!

**Comment by Saduf2019** [comment]
I face a different error on tf-nigtly [2.4.0-dev20201021], please find the [gist here](https://colab.research.google.com/gist/Saduf2019/996eb8311c72a0aaeb9f2258f600219e/untitled450.ipynb).

**Comment by sushreebarsa** [comment]
I faced a different error on TF 2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/2bea00ac27ed38c70b7e62f8666141b8/untitled450.ipynb#scrollTo=0t1dEzszdUWS) ..Thanks!

**Comment by rxsang** [comment]
The error message indicates your client and TPU worker are not the same version. Can you make sure they are both in TF nightly or TF 2.5?

**Comment by sushreebarsa** [comment]
@rxsang thanks for noticing that ,I have updated the gist in TF v2.5 .

**Comment by tilakrayal** [comment]
@saahiluppal,
You can iterate over the entire input train_dist_dataset inside a [tf.function](https://www.tensorflow.org/api_docs/python/tf/function) using the for x in ... construct or by creating iterators like you did above. 

Please take a look at below example  which helps in wrapping one epoch of training with a [@tf.function](https://www.tensorflow.org/api_docs/python/tf/function) decorator and iterating over train_dist_dataset inside the function.

```python
@tf.function
def distributed_train_epoch(dataset):
  total_loss = 0.0
  num_batches = 0
  for x in dataset:
    per_replica_losses = strategy.run(train_step, args=(x,))
    total_loss += strategy.reduce(
      tf.distribute.ReduceOp.SUM, per_replica_losses, axis=None)
    num_batches += 1
  return total_loss / tf.cast(num_batches, dtype=tf.float32)

for epoch in range(EPOCHS):
  train_loss = distributed_train_epoch(train_dist_dataset)

  template = ("Epoch {}, Loss: {}, Accuracy: {}")
  print(template.format(epoch + 1, train_loss, train_accuracy.result() * 100))

  train_accuracy.reset_states()
```

Please refer to the tutorial https://www.tensorflow.org/tutorials/distribute/custom_training and replace the **MirroredStrategy** in the tutorial with **TPUStrategy** ( initialization etc associated wrt TPUStrategy). Thank you!
